src/integer_mod_q/mat_zq/solve.rs

```diff
@@ -58,44 +62,314 @@ impl MatZq {
+                        .unwrap();
+
+                    permutation.swap(row_count.try_into().unwrap(), row_nr.try_into().unwrap());
+                }
+
+                // Save the position of `1` for that row.
+                indices.push((row_count, column_nr));
+                row_count += 1;
+            // Search for a not invertible entry to divide the modulus.
+            } else if let Some(row_nr) =
+                find_not_invertible_entry_column(&matrix_identity_base_gauss, column_nr, used_rows)
+            {
+                // Factorize the modulus with the found entry, compute solutions
+                // for the system under the split modulus and use the
+                // Chinese Remainder Theorem to compute a solution based on these
+                // sub-solutions.
+                let entry: Z = matrix_identity_base_gauss
+                    .get_entry(row_nr, column_nr)
+                    .ok()
+                    .unwrap();
+                self.factorization_and_crt(y, entry);
+            }
         }
 
-        // check if we have computed a solution and return `None`
-        if &(self * &out) != y {
+        // Return `None` if the matrix has no full rank.
+        if indices.is_empty()
+            || indices[indices.len() - 1].0 + 1 < matrix_identity_base_gauss.get_num_rows()
+        {
             return None;
         }
 
+        // Pick the first columns out of the original matrix that form an invertible matrix.
+        // Those columns exist, since the matrix was checked to be full rank.
+        let mut square_matrix = MatZq::new(
+            matrix_identity_base_gauss.get_num_rows(),
+            matrix_identity_base_gauss.get_num_rows(),
+            self.get_mod(),
+        );
+        for (current_column, (_row_nr, column_nr)) in indices.iter().enumerate() {
+            square_matrix
+                .set_column(current_column, self, *column_nr)
+                .unwrap();
+        }
+
+        // The inverse of the previously picked square matrix consists of the last
+        // columns of `matrix_identity_base_gauss`, since we concatenated an identity matrix.
+        let matrix_identity_gauss =
+            MatZq::from((&MatZ::from(&matrix_identity_base_gauss), &self.get_mod()));
+        let mut matrix_base_inv = MatZq::new(
+            matrix_identity_gauss.get_num_rows(),
+            matrix_identity_gauss.get_num_rows(),
+            &self.get_mod(),
+        );
+        for row_nr in 0..matrix_identity_gauss.get_num_rows() {
+            matrix_base_inv
+                .set_column(
+                    row_nr,
+                    &matrix_identity_gauss,
+                    row_nr + self.get_num_columns(),
+                )
+                .unwrap();
+        }
+
+        // Use the method from "Exact Solution of Linear Equations Using P-Adic Expansions"
+        // <https://link.springer.com/content/pdf/10.1007/BF01459082.pdf/>
```

Comment:
please include the reference in the format we used for the rest of our implementation. e.g. use the format we use in the crpto crate for g-trapdoors